### PR TITLE
Build man pages on demand only

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: man/html.in [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+if [ -z "$1" ]; then
+    echo "Use: $0 page-name (with no section suffix)"
+    exit 1
+fi
+
+target="man/$1.html"
+ninja -C "@BUILD_ROOT@" "$target"
+set -x
+xdg-open build/"$target"
```
Comment (poettering) [inline man/html.in]: why not `exec`?

File: man/man.in [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    exit 1
+fi
+
+target=$(ninja -C "@BUILD_ROOT@" -t query man/man | grep -E -m1 "man/$1\.[0-9]$" | awk '{print $2}')
+if [ -z "$target" ]; then
+    echo "Cannot find page $1"
+    exit 1
+fi
+ninja -C "@BUILD_ROOT@" "$target"
+man build/"$target"
```
Comment (poettering) [inline man/man.in]: why no `exec`?

File: man/html.in (hunk 1 above)
Comment (keszybz) [reply man/html.in]: Will add.

**Comment by keszybz** [comment]
Updated. ubuntuautopkgtests are configured somewhere externally, and the build will fail until that configuration is changed.

@xnox ?

**Comment by xnox** [comment]
Pushed https://salsa.debian.org/systemd-team/systemd/commit/0a67c4bc158dc6af83173558c79790bfa41f30ca

Please retrigger tests? Not sure how to do it myself, push an empty commit / rebase?

**Comment by xnox** [comment]
Also wonder if we can switch to an experimental branch somehow, as master should be frozen for packaging changes like these.

**Comment by keszybz** [comment]
@xnox thanks, this seems to have worked.

Only bionic-arm64 failed: " Process 923 (bash) of user 0 dumped core."
I'll restart it.

**Comment by evverx** [comment]
"Process 923 (bash) of user 0 dumped core" is kind of expected because there is a test crashing bash on purpose to see how `systemd-coredump` fares. bionic-arm64 fails from time to time due to https://github.com/systemd/systemd/issues/11104. I just forgot to turn it off after I turned it on in https://github.com/systemd/systemd/pull/12265#issuecomment-481714473. I'll turn it off again shortly.

**Comment by evverx** [comment]
The webhook is now inactive.

By the way, it would probably make sense to mention that in `NEWS`. Though, to be honest, I think most people kind of expect the documentation to be built by default and it shouldn't be turned off to save some time for a couple of developers.

**Comment by poettering** [comment]
hmm, yeah, i wonder if it wouldn't be better to leave the man pages on by default, but make it something that is easily turned off...

**Comment by evverx** [comment]
I think that passing `-Dman=false` explicitly is already a relatively easy way to turn it off. Maybe to make it easier to not forget to pass it, we could put a bash script in, say, `tools/` running something like
```
meson -Dman=false ./build
ninja -C ./build
```
(which also will help to save a few keystrokes :-))

**Comment by keszybz** [comment]
Nah. I know that changing the default required changes in the build scripts, but we only needed to do this once and it's already done. OTOH, one-off builds by developers happen all the time... Every time everybody clones our repo to test something... I think we need to make this as quick and easy as possible.

**Comment by evverx** [comment]
> I think we need to make this as quick and easy as possible.

That's a noble goal but why do you think people want to ditch the documentation? My assumption (based of the fact that it's basically the first time I've heard anyone complaining about how long it takes to build systemd with the documentation) is that nobody (apart from a few people who are capable of passing `-Dman=false) cares about that.

**Comment by evverx** [comment]
Just to be clear, I'm not saying it's going to be hard to pass `-Dman=true` if this is merged. It's just it's unclear to me why the documentation is treated as a burden (when in fact it's as important as the code). Following this logic it would make sense to also turn off all the tests. Who needs them anyway? :-)

**Comment by keszybz** [comment]
Man pages are not like code, because they are almost fully independent. For code, you want to build everything to make sure stuff still compiles and there are no warnings. For man pages, 99% of the time, when working on man pages it is only useful and necessary to build and preview the man page that you are just modifying. And when doing changes to the code base, man page rebuilds are just a  pointless drag.

Building the docs is maybe half of the build time, and docs get rebuilt when configuration changes, so for example bisects are unnecessarily slow.
When using docs from the build tree, one needs to invoke man with the build path anyway (to get the freshly build page), which means that one needs to know the section number or build all the pages first and use tab completion. The script which is added in this patch simplifies this, because it "guesses" the section number. Once we have the script, it turns out that building the man page on demand is quick enough that it's just as convenient to only build them on demand.

We don't build html pages by default, even though we could. We only build them when explicitly configured to do so, or on demand for doc-sync.

This (as many other things) comes down to a trade-off. I think that making developer experience and bisects and such more pleasant in this case outweighs the need to set an option when doing packaging. In that latter case, it is good practice to set the option explicitly anyway.

**Comment by evverx** [comment]
> Man pages are not like code, because they are almost fully independent. For code, you want to build everything to make sure stuff still compiles and there are no warnings. For man pages, 99% of the time, when working on man pages it is only useful and necessary to build and preview the man page that you are just modifying. And when doing changes to the code base, man page rebuilds are just a pointless drag.

That's correct in terms of build systems but my assumption is that people who build the code in general want to also build the documentation because, well, it's hard to deal with `systemd` (and especially change its code) without reading the documentation (and it doesn't matter whether they change any manpages or not).

Oh, almost forgot, @mrc0mmand could you pass `-Dman=true` on CentOS explicitly as well?

**Comment by evverx** [comment]
Regarding `git bisect`, it's been a rather unpleasant experience since the switch from `autotools` to `meson` :-) Plus the shenanigans with `missing.h` have made it impossible to do anything without disabling `systemd-networkd`, which simply won't compile if you go backward far enough and even turning off `systemd-networkd` doesn't always work. So I don't think `-Dman=false` by default will help much in this regard.

**Comment by yuwata** [comment]
LGTM. And I like this. As I always disable `-Dman` and enable it when necessary. But this must be mentioned in NEWS.

**Pull request merged**